### DISCUSSION TITLE
External_relay hostname strategy and ratelimit whitelisted_ip

### DISCUSSION THREAD

**Discussion by izydorczyk** [rspamd/rspamd#4771] [Q&A] [answered]
Is it possible to disable skipping received headers for the **ratelimit** module, specifically for the **whitelisted_ip** setting?

According to our testing, the selector "ip" uses IP addresses from `received` headers affected by the external_relay plugin.
If we have this:
ratelimit
```
whitelisted_ip = ["/etc/rspamd/local.d/maps.d/ratelimit_whitelisted_ip.map"]
rates = {
  from_ip = {
    selector = 'ip';
    bucket = {
      burst = 1000;
      rate = "6 / 1m"; # leak 6 messages per minute
    }
  }
}
``` 
and external_relay
```
rules {
  EXTERNAL_RELAY_AUTHENTICATED {
    strategy = "hostname_map";
    hostname_map =  "/etc/rspamd/local.d/maps.d/external_relay_hostname.map";
  }
}
```
then buckets are created based on the IP addresses of the "original" senders (indirect) instead of the actual (direct) MTAs.

To illustrate the mail message path:
Source: External MTA --> MTA1 --> MTA2 --> MTA1 --> Destination: Microsoft365 

MTA1 Rspamd will use the bucket for IP of External MTA even in the case where the same message returns from MTA2 back to MTA1. This is even when MTA2 is whitelisted in the ratelimit module.

Can the whitelisting of the ratelimit plugin be tweaked to ignore the external_relay plugin?
Or is there another solution, like a bucket for when the message returns from a known (specific) MTA that has a more pass-through configuration?

**Comment by fatalbanana** [comment, marked as answer]
You could disable `ratelimit` module based on the presence of `EXTERNAL_RELAY_AUTHENTICATED` symbol:

~~~lua
-- /etc/rspamd/rspamd.local.lua

rspamd_config:add_condition('RATELIMIT_CHECK', function(task)
  return not task:has_symbol('EXTERNAL_RELAY_AUTHENTICATED')
end)
~~~

**Reply by izydorczyk** [reply]
thx
